MRG[+1]: adds limo dataset example code

JoseAlanis opened the pull request:
#### Reference issue
Closes #6075 

#### What does this implement/fix?
Add example code to load single subject data from the LIMO dataset, creating an MNE epochs structure and computing the evoked response for experimental conditions.

larsoner commented on examples/datasets/load_limo_dataset.py:
```diff
@@ -0,0 +1,173 @@
+
+The code allows to:
+
+    - Fetch single subjects epochs data for the LIMO data set.
```
do not indent this, it will show up as a quoted list instead of standard list

JoseAlanis replied:
Ok, I hope the new version is better.

larsoner commented on examples/datasets/load_limo_dataset.py:
```diff
@@ -0,0 +1,173 @@
+    - Plot the evoked response for experimental conditions.
+
+.. note:: This example assumes that the LIMO data set has already
+          been downloaded and stored in home directory.
```
We should provide a proper `mne.datasets.limo.data_path()` function for this

agramfort commented:
+1 to add a proper data fetcher

>

dengemann commented:
> +1 to add a proper data fetcher

Good @agramfort @larsoner . This is what I wanted to here. @JoseAlanis let's do a limo data fetcher just like for the other packages. 

We should then do a data loader like:  ```mne.datasets.eegbci.load_data```

jona-sassenhagen commented on examples/datasets/load_limo_dataset.py:
```diff
@@ -0,0 +1,173 @@
+    events[:, 0] = list(range(len(design)))
+    events[:, 2] = design[:, 1]
+    # event ids, such that B == 1
+    event_id = dict(A=0, B=1)
```
Let's use decent condition names here.

JoseAlanis replied:
"Face A" and "Face B" that's pretty much how they call them in the LIMO paper.

jona-sassenhagen commented on examples/datasets/load_limo_dataset.py:
```diff
@@ -0,0 +1,173 @@
+    # add external electrodes (e.g., eogs)
+    ch_names = montage.ch_names[:-3] + ['EXG1', 'EXG2', 'EXG3', 'EXG4']
+    # match individual labels to labels in montage
+    found_inds = [ii for ii, ee in enumerate(ch_names) if ee in labels]
```
ee => ch_name

jona-sassenhagen commented on examples/datasets/load_limo_dataset.py:
```diff
@@ -0,0 +1,173 @@
+        # avoid copy when fancy indexing.
+        all_data[:, target, :] = obs_data[:, source, :]
+    obs_data = all_data
+    obs_data /= 1e6  # data to V (to match MNE)
```
obs_data = all_data / 1e6

jona-sassenhagen commented on examples/datasets/load_limo_dataset.py:
```diff
@@ -0,0 +1,173 @@
+    types = []
+    for ch in ch_names:
+        types.append(
+            'eog' if ch in ('EXG1', 'EXG2', 'EXG3', 'EXG4') else 'eeg')
```
List comprehension.

`types = ["eog" if ch.startswith("EXG") else "eeg" for ch in ch_names]`

jona-sassenhagen commented on examples/datasets/load_limo_dataset.py:
```diff
@@ -0,0 +1,173 @@
+# --- Use load_limo_epochs() function to fetch subject data ---
+
+# fetch data from subject 1
+data, design = load_limo_epochs(subject = 1, path=None)
```
No spaces for parameters in functions (`subject=1`)

jona-sassenhagen commented on examples/datasets/load_limo_dataset.py:
```diff
@@ -0,0 +1,173 @@
+data.info
+
+# plot evoked response for condition A
+data['A'].average().plot_joint(times=[.09, .15])
```
If the fetcher is put into a proper function, the actual example can be more interesting.

Look at the examples for e.g. the SPM data.

jona-sassenhagen commented on examples/datasets/load_limo_dataset.py:
```diff
@@ -0,0 +1,173 @@
+    # create events matrix
+    events = np.zeros(shape=(len(design), 3), dtype=int)
+    events[:, 0] = list(range(len(design)))
+    events[:, 2] = design[:, 1]
```
It might actually be more economic to just construct it properly in the first place:

`np.array([list(range(len(design))), np.zeros(len(design)), design[:, 1]])`

Or something like that.

jona-sassenhagen commented on examples/datasets/load_limo_dataset.py:
```diff
@@ -0,0 +1,173 @@
+
+
+# import necessary packages
+import os.path as op
```
Blank line here (core imports)

jona-sassenhagen replied:
@JoseAlanis resolve conversations when you've done so.

jona-sassenhagen commented on examples/datasets/load_limo_dataset.py:
```diff
@@ -0,0 +1,173 @@
+        The path to store the downloaded data. Defaults to home directory.
+    subject : int | str
+        Subject to use. Can be in the range from 2 to 18.
+        If string, must be 'S1', 'S2' etc.
```
Actually, do we really want to support both str and int? Did we discuss this?

JoseAlanis replied:
No, I don't think so. I left it as it is for now. But I can change the input parameters if you guys think it's better.

jona-sassenhagen commented on examples/datasets/load_limo_dataset.py:
```diff
@@ -0,0 +1,173 @@
+    Parameters
+    ----------
+    path : str
+        The path to store the downloaded data. Defaults to home directory.
```
Order of docstring must match order of call signature, i.e. subject first.

jona-sassenhagen commented on examples/datasets/load_limo_dataset.py:
```diff
@@ -0,0 +1,173 @@
+
+    # -- 7) interpolate missing channels
+    if interpolate is True and montage is not None:
+        # code bad channels
```
"code" bad channels?

JoseAlanis replied:
Sorry, I don’t know what I was trying to say there. I deleted the comment.

jona-sassenhagen commented:
@JoseAlanis any progress?

JoseAlanis commented:
Hey guys, yes, I added some pieces of code I've been working on. 
I added the functions ``mne.datasets.limo.data_path()`` and ``mne.datasets.limo.load_data()`` as suggested. Looking forward to hear your comments and suggestions.

The functions handle the downloading and fetching of the LIMO epochs more in the style of the other datasest that are already integrated in MNE. One difficulty I came upon however is that the LIMO files are stored in one big .zip folder (~8 GB) online. So the first downloading and extracting of the files takes time and cover a big chunk of space on disk. Also, there are some nested .zips and .tars in the download file. Right now I just loop through the levels and unpack them, but it would be nice to hear if there is a better way of doing this.

The example code in ``examples/datasets/`` is not quite ready jet. I'll add some more analyses soon.

jona-sassenhagen commented on mne/datasets/limo/limo.py:
```diff
@@ -0,0 +1,253 @@
+import re
+import zipfile
+import tarfile
+from sys import stdout
```
Group all standardlib imports together. Then scientific python imports. Then MNE imports.

jona-sassenhagen replied:
Group all standardlib imports together. Then scientific python imports. Then MNE imports.

jona-sassenhagen commented on mne/datasets/limo/limo.py:
```diff
@@ -0,0 +1,253 @@
+import scipy.io
+import numpy as np
+import pandas as pd
+import mne
```
We typically don't import mne. We just import, as you do below, from relative paths.

jona-sassenhagen replied:
We typically don't import mne. We just import, as you do below, from relative paths.

jona-sassenhagen commented on mne/datasets/limo/limo.py:
```diff
@@ -0,0 +1,253 @@
+    limo_dir = op.split(destination)[0]
+
+    # check if zip.-folders are contained
+    is_zip = True if '.zip' in [x.lower() for x in op.splitext(destination)] else False
```
`is_zip = any(x.endswith(".zip") for x in op.splittext(destination))`

Would this work?

JoseAlanis replied:
yup, it does. I'm now using the ``.endswith("")`` method throughout the code.

jona-sassenhagen commented on mne/datasets/limo/limo.py:
```diff
@@ -0,0 +1,253 @@
+        if not op.isdir(op.dirname(destination)):
+            os.makedirs(op.dirname(destination))
+        _fetch_file(url, destination, print_destination=False)
+        do_unzip = True
```
```
do_unzip = not op.isfile(destination) or force_update
if do_unzip:
    if op.isfile...
```
Etc

jona-sassenhagen commented on mne/datasets/limo/limo.py:
```diff
@@ -0,0 +1,253 @@
+
+    # decompress if necessary
+    if is_zip and do_unzip:
+        z1 = zipfile.ZipFile(destination)
```
Does zipfile work as a context manager (`with zipfile.ZipFile(destination) as z1`)? If yes, please use it

JoseAlanis replied:
yes. I'm using ``zipfiel.Zipfile()`` and ``tarfile.open()`` as a context manager. I changed the while donwload-zip-thing a bit though, since we only really need to extract files during first time download or ``force_update=True``.

jona-sassenhagen commented on mne/datasets/limo/limo.py:
```diff
@@ -0,0 +1,253 @@
+        z1.close()
+        # list files in .zip-folder and look for further .zip-folders to decompress
+        extracted_files = [op.join(limo_dir, f) for f in files]
+        sub_files = [e for e in extracted_files if ".zip" in e or ".tar" in e]
```
use `e.endswith`

jona-sassenhagen commented on mne/datasets/limo/limo.py:
```diff
@@ -0,0 +1,253 @@
+        if sub_files:
+            stdout.write('Decompressing the extracted files ...\n')
+            for i in sub_files:
+                if '.zip' in i:
```
same

jona-sassenhagen commented on mne/datasets/limo/limo.py:
```diff
@@ -0,0 +1,253 @@
+
+        if sub_files:
+            stdout.write('Decompressing the extracted files ...\n')
+            for i in sub_files:
```
let's not use single-character names

jona-sassenhagen commented on mne/datasets/limo/limo.py:
```diff
@@ -0,0 +1,253 @@
+                    z2.extractall(limo_dir)
+                    z2.close()
+                    # continue decompressing if necessary
+                    sub_sub_files = [j for j in os.listdir(sub_dir) if '.zip' in j]
```
same, same

jona-sassenhagen commented on mne/datasets/limo/limo.py:
```diff
@@ -0,0 +1,253 @@
+                    sub_dir = op.splitext(i)[0]
+                    z2 = zipfile.ZipFile(i)
+                    z2.extractall(limo_dir)
+                    z2.close()
```
Yes, please use the context manager.

jona-sassenhagen commented on mne/datasets/limo/limo.py:
```diff
@@ -0,0 +1,253 @@
+                        z3.close()
+                elif '.tar' in i:
+                    t = tarfile.open(os.path.join(limo_dir, i))
+                    t.extractall(limo_dir)
```
Does it need to be closed?
Context manager?

jona-sassenhagen commented on mne/datasets/limo/limo.py:
```diff
@@ -0,0 +1,253 @@
+                        z3 = zipfile.ZipFile(os.path.join(sub_dir, k))
+                        z3.extractall(sub_dir)
+                        z3.close()
+                elif '.tar' in i:
```
Under what conditions is it tar vs. zip?

I think here it shows more clearly why more explicit names are better. What is `i`? :)

jona-sassenhagen commented on mne/datasets/limo/limo.py:
```diff
@@ -0,0 +1,253 @@
+        The epochs.
+    """  # noqa: E501
+    if url is None:
+        url = 'http://datashare.is.ed.ac.uk/download/DS_10283_2189.zip'
```
I personally prefer ` url = 'http://datashare.is.ed.ac.uk/download/DS_10283_2189.zip' if url is None else url`

jona-sassenhagen replied:
Might not actually be better in this case.

JoseAlanis replied:
I thought this would make the statement more explicit. I'll leave it as it is for now, but i'm fine with your suggestion if you think it's better.

jona-sassenhagen commented on mne/datasets/limo/limo.py:
```diff
@@ -0,0 +1,253 @@
+    elif isinstance(subject, str):
+        if not subject.startswith('S'):
+            raise ValueError('`subject` must start with `S`')
+        subj = subject
```
Can we do a check here for if this subject actually exists?

JoseAlanis replied:
yup, I changed the settings to only accept ``int`` in range 1-18 as subject identifier (we've thought about this before).

jona-sassenhagen commented on mne/datasets/limo/limo.py:
```diff
@@ -0,0 +1,253 @@
+
+    # -- 2) get epochs information from structure
+    # sampling rate
+    sfreq = data_info['LIMO']['data'][0][0][0][0]['sampling_rate'][0][0]
```
`data_info['LIMO']['data'][0][0][0][0]` is used 3+ times right? Maybe give it a name.

jona-sassenhagen commented on mne/datasets/limo/limo.py:
```diff
@@ -0,0 +1,253 @@
+                                    np.zeros(len(design)),
+                                    design[:, 1]]).astype(int))
+    # event ids, such that Face B == 1
+    event_id = {'Face A': 0, 'Face B': 1}
```
I'm now thinking we should actually call them "Face/A", "Face/B" now.

jona-sassenhagen commented on mne/datasets/limo/limo.py:
```diff
@@ -0,0 +1,253 @@
+    # -- 3) extract channel labels from LIMO structure
+    # get individual labels
+    labels = data_info['LIMO']['data'][0][0][0][0]['chanlocs']['labels']
+    labels = [ll[0] for ll in labels[0]]
```
Or `[label for label, *_ in labels]`

jona-sassenhagen commented on mne/datasets/limo/limo.py:
```diff
@@ -0,0 +1,253 @@
+    ch_names = montage.ch_names[:-3] + ['EXG1', 'EXG2', 'EXG3', 'EXG4']
+    # match individual labels to labels in montage
+    found_inds = [ii for ii, ch_name in enumerate(ch_names) if ch_name in labels]
+    missing_chans = [ch for ch in ch_names if ch not in labels]
```
Call it ch_name both times. (Or ch, but better ch_na,es)

jona-sassenhagen commented on mne/datasets/limo/limo.py:
```diff
@@ -0,0 +1,253 @@
+    info = mne.create_info(ch_names=ch_names, ch_types=types, sfreq=sfreq)
+
+    # get faces and noise variables from design matrix
+    faces = ['A' if list(events[:, 2])[i] == 0 else 'B' for i in list(events[:, 2])]
```
You're constructing `list(events[:, 2])` quite often here. Construct it outside the list comp.

jona-sassenhagen replied:
It becomes a lot nicer if you do!

jona-sassenhagen commented on mne/datasets/limo/limo.py:
```diff
@@ -0,0 +1,253 @@
+    noise = list(design[:, 2])
+    # create epochs metadata
+    metadata = {'Face': faces,
+                'Noise': noise}
```
One line

jona-sassenhagen commented on mne/datasets/limo/limo.py:
```diff
@@ -0,0 +1,253 @@
+
+    # get faces and noise variables from design matrix
+    faces = ['A' if list(events[:, 2])[i] == 0 else 'B' for i in list(events[:, 2])]
+    noise = list(design[:, 2])
```
You are doing it here!!!

jona-sassenhagen commented on mne/datasets/limo/limo.py:
```diff
@@ -0,0 +1,253 @@
+
+    # -- 6) Create custom epochs array
+    epochs = mne.EpochsArray(obs_data, info, events, tmin, event_id, metadata=metadata)
+    epochs.set_montage(montage=montage)
```
create_info takes a montage arg, can't you use it there?

jona-sassenhagen commented on mne/datasets/limo/limo.py:
```diff
@@ -0,0 +1,253 @@
+    epochs.info['bads'] = missing_chans  # missing channels are marked as bad.
+
+    # -- 7) interpolate missing channels
+    if interpolate is True and montage is not None:
```
`if interpolate and montage:` might be enough. *Might*.

JoseAlanis replied:
I change this to ``if interpolate:``, since im forcing the setting of the montage above.

jona-sassenhagen commented on mne/datasets/limo/limo.py:
```diff
@@ -0,0 +1,253 @@
+    # create events matrix
+    events = np.transpose(np.array([list(range(len(design))),
+                                    np.zeros(len(design)),
+                                    design[:, 1]]).astype(int))
```
Just arr.T instead of np.transpose(arr).

jona-sassenhagen commented on examples/datasets/load_limo_dataset.py:
```diff
@@ -0,0 +1,64 @@
+limo_epochs.info
+
+# look at metadata
+limo_epochs.metadata
```
Hm? If this is not printed, we don't actually see it right? Also .head is probably better.

jona-sassenhagen commented:
Cool, thanks @JoseAlanis ! I left a few comments to begin with.

jona-sassenhagen commented:
>The functions handle the downloading and fetching of the LIMO epochs more in the style of the other datasest that are already integrated in MNE. One difficulty I came upon however is that the LIMO files are stored in one big .zip folder (~8 GB) online. So the first downloading and extracting of the files takes time and cover a big chunk of space on disk. Also, there are some nested .zips and .tars in the download file. Right now I just loop through the levels and unpack them, but it would be nice to hear if there is a better way of doing this.

This is the big question.
Downloading 20 GB is scary.

dengemann commented:
@jona-sassenhagen let me know when you're done. Then I'll start having a look.

jona-sassenhagen commented:
@JoseAlanis can you resolve the issues I raised?

Afterwards can you ping @dengemann and ask him to give it a pass? Thanks

JoseAlanis commented:
Alright, I integrated your suggestions, @jona-sassenhagen. What do you think, @dengemann?

larsoner commented:
LGTM shows an error, as do the `make pep` tests:

https://travis-ci.org/mne-tools/mne-python/jobs/532166182#L3433-L3457

JoseAlanis commented:
ok, I changed the code a little bit, hope the changes take care of the issues.

agramfort commented:
example files should be called plot_
you need to update the CI files to this dataset is downloaded

do a `git grep kiloword` and see where new datasets should be list / documented

jona-sassenhagen commented:
@JoseAlanis check Travis:

https://travis-ci.org/mne-tools/mne-python/jobs/533895775#L2659

You need to restrict the pandas import in _requires_pandas or something like that. See if you can find how yourself, otherwise ask again.

jona-sassenhagen commented:
Also check here: https://travis-ci.org/mne-tools/mne-python/jobs/533895776#L2810

JoseAlanis commented:
Alright, thanks, @jona-sassenhagen. Hope I got it right

agramfort commented:
rendered example
https://13245-1301584-gh.circle-artifacts.com/0/dev/auto_examples/datasets/plot_limo_data.html

>

dengemann commented:
the face effect looks really aweful, especially compared to the average. Any clue what’s going on here?

> On 21 May 2019, at 09:58, Alexandre Gramfort <notifications@github.com> wrote:
> 
> rendered example
> https://13245-1301584-gh.circle-artifacts.com/0/dev/auto_examples/datasets/plot_limo_data.html
> 
> >
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

jona-sassenhagen commented:
Do we need an interaction term here?

jona-sassenhagen commented on mne/datasets/limo/limo.py:
```diff
@@ -0,0 +1,232 @@
+    # tmin and tmax
+    tmin = data_info['start'][0][0]
+    # create events matrix
+    sample = list(range(len(design)))
```
`np.arange`?

jona-sassenhagen commented on mne/datasets/limo/limo.py:
```diff
@@ -0,0 +1,232 @@
+    info = create_info(ch_names, sfreq, types, montage)
+    # get faces and noise variables from design matrix
+    event_list = list(events[:, 2])
+    faces = ['A' if event == 0 else 'B' for event in event_list]
```
Could even lose the `== 0` as 0 is False right?

JoseAlanis replied:
True, thank for the hint.

JoseAlanis commented on examples/datasets/plot_limo_data.py:
```diff
@@ -0,0 +1,80 @@
+design = design.assign(Intercept=1)
+names = ['Intercept', 'Face_Effect', 'Noise']
+# check design matrix
+design[names].head()
```
true, I'm explicitly printing the metadata now

JoseAlanis commented on examples/datasets/plot_limo_data.py:
```diff
@@ -0,0 +1,80 @@
+design = limo_epochs.metadata
+design['Face_Effect'] = np.where(design['Face'] == 'A', 1, 0)
+design = design.assign(Intercept=1)
+names = ['Intercept', 'Face_Effect', 'Noise']
```
alright, I adapted the names of the columns in the design matrix to go with my explanation the in the example. Hope it's better now.

JoseAlanis commented on examples/datasets/plot_limo_data.py:
```diff
@@ -0,0 +1,80 @@
+reg['Noise'].beta.plot_joint(title='Effect of Noise',
+                             ts_args=dict(time_unit='s'),
+                             topomap_args=dict(time_unit='s'),
+                             times=[.125, .225])
```
yup, I guess, I deleted it.

JoseAlanis commented:
Thanks for the comments @jona-sassenhagen. Like we discusses earlier with @dengemann, I added a more extensive explanation of the LIMO-paradigm and the effects that can be expected from linear regression. Hope the example is clearer now. Specially the reason as to why some effects look nice and others just don't show any systematic differences.
I'm leaving the interaction term out for now and just going with the main effect of phase-coherence and face condition, as these are the effects that are discussed more closely in the original LIMO paper.
Hope the example renders well with the changes made.
What do you think?

dengemann commented:
Is the CI error related to the PR?

larsoner commented on mne/datasets/limo/limo.py:
```diff
@@ -0,0 +1,232 @@
+    return limo_dir
+
+
+@requires_pandas
```
This error on Travis is real:
```
Found un-nested imports for:
E         pandas
```
We don't use `requires_` decorators for functions in the module, just for tests. See how we check for `pandas` in other functionality, such as `to_data_frame`

JoseAlanis replied:
ok, thanks. If I understand correctly, one solution could be to add ``from ...utils import  _check_pandas_installed`` and check for pandas inside the function by calling ``pd = _check_pandas_installed()``, insted of importing pandas explicitly?
Would I keep the ``requires_pandas`` decorator for testing purposes or should I delete it?

larsoner replied:
> ... calling pd = _check_pandas_installed(), insted of importing pandas explicitly?

Yes

> Would I keep the requires_pandas decorator for testing purposes or should I delete it?

Delete it from within the `mne` module but decorate any test that requires it, if there are any. From a quick look at the code I don't see any, which is normal/fine for datasets. CircleCI ends up implicitly testing the dataset functionality when building examples.

JoseAlanis replied:
ok, many thanks for the explanation.

dengemann commented:
Where are we here? Ping ping

JoseAlanis commented:
I took care of the errors. Something else I can do to get the PR going?

dengemann commented:
Ok great! Looks good to me. I think we can live with some coverage issues for now. If no one objects, I'll merge later.


FYI @GRousselet @CPernet MNE about to feature LIMO :-) Feedback is welcome.

agramfort commented on mne/datasets/limo/limo.py:
```diff
@@ -0,0 +1,231 @@
+
+
+@verbose
+def load_data(subject, path=None, interpolate=False, force_update=False,
```
I would not put the interpolation logic in the loader

JoseAlanis replied:
Ok, I guess we can remove the interpolation and let people interpolate missing channels  (currently marked as "bad") later if desired. Would that be a better solution?

dengemann replied:
At least we make no API commitment that way and don't lose much by pulling it out.

agramfort replied:
Yes. Remove the option

JoseAlanis replied:
Ok. I removed the option in the loader and added a comment in ``plot_limo_data`` regarding interpolation of missing channels.

agramfort commented:
Does this fetch the full limo dataset?
Does the example show something from a paper?

dengemann commented:
> Does this fetch the full limo dataset?

no, just the EEG of one subject

> Does the example show something from a paper?

Yes it shows results compatible with the LIMO paper, however not the group analysis

dengemann commented:
Yes! Absolutely.

> On 31 May 2019, at 11:22, José C. García Alanis <notifications@github.com> wrote:
> 
> @JoseAlanis commented on this pull request.
> 
> In mne/datasets/limo/limo.py:
> 
> > +                files = [op.join(limo_dir, file) for file in z1.namelist()]
> +                stdout.write('Decompressing %g files from\n'
> +                             '"%s" ...' % (len(files), destination))
> +                z1.extractall(op.join(limo_dir, subject_id))
> +                stdout.write(' [done]\n')
> +                z1.close()
> +                os.remove(destination)
> +
> +    # update path if desired
> +    _do_path_update(path, update_path, key, name)
> +
> +    return limo_dir
> +
> +
> +@verbose
> +def load_data(subject, path=None, interpolate=False, force_update=False,
> Ok. I removed the option in the loader and added a comment in plot_limo_data regarding interpolation of missing channels.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

dengemann commented:
Ok time to merge 🚀 - thank you @JoseAlanis + reviewers!

larsoner commented:
This broke CircleCI:

https://circleci.com/gh/mne-tools/mne-python/13452

 Is there a user prompt for the dataset? If so you need to do what we do for the brainstorm data. Also need to modify mne.datasets.utils._download_all_example_data, which Circle uses on full builds.

Can someone look ASAP?

massich commented:
This was not **squashed**

jona-sassenhagen commented:
Denis :D
